lib/webauthn/authentication_service.rb

```diff
@@ -2,21 +2,25 @@
 require "cose"
 
 module DiscourseWebauthn
-  class SecurityKeyAuthenticationService < SecurityKeyBaseValidationService
+  class AuthenticationService < BaseValidationService
     ##
     # See https://w3c.github.io/webauthn/#sctn-verifying-assertion for
     # the steps followed here. Memoized methods are called in their
     # place in the step flow to make the process clearer.
     def authenticate_security_key
       if @params.blank? || (!@params.is_a?(Hash) && !@params.is_a?(ActionController::Parameters))
-        return false
+        raise(
+          MalformedPublicKeyCredentialError,
+          I18n.t("webauthn.validation.malformed_public_key_credential_error"),
+        )
       end
 
       # 3. Identify the user being authenticated and verify that this user is the
       #    owner of the public key credential source credentialSource identified by credential.id:
       security_key = UserSecurityKey.find_by(credential_id: @params[:credentialId])
-      raise(NotFoundError, I18n.t("webauthn.validation.not_found_error")) if security_key.blank?
-      if security_key.user != @current_user
+      raise(KeyNotFoundError, I18n.t("webauthn.validation.not_found_error")) if security_key.blank?
+      if @factor_type == UserSecurityKey.factor_types[:second_factor] &&
+           security_key.user != @current_user
```

Comment:
Is there a possibility that `@current_user` can be `nil` here? I think when first factor is used, `@current_user` will be `nil` and there might be a chance we compare `nil == nil` if `security_key.user` returns `nil`?

Reply:
Yes, `@current_user` will be `nil` for first factor. In theory, we're only running this check for second factor keys, and all second factor keys _should_ belong to a user. Would it be safer to add an additional check here like `if @current_user != nil`? I'm not against it.

Reply:
It looks like [we don't clean up `security_keys`](https://github.com/discourse/discourse/blob/b267c1054d4065c264ab30fb380574e2ee3e1047/app/models/user.rb#L98-L99) when a user record is deleted so there is a chance that `security_key.user` might be `nil`. 

Since this is security related, I think it is better to ensure that `security_key.user` and `@current_user` are both not `nil` before making the comparison.